### ISSUE TITLE
Invalid repository discovery when workspace .git dir is a symlink/junction

### ISSUE THREAD

**Issue by AnrDaemon** [git-for-windows/git#4688]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.42.0.windows.2
cpu: x86_64
built from commit: 2f819d1670fff9a1818f63b6722e9959405378e3
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon

```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19045.3570]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
> type "$env:USERPROFILE\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Portable install from .tar.bz2

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

```
$ ls -ld .git
lrwxrwxrwx 1 anrdaemon 197121 45 ноя 14 14:27 .git -> ../invapi-php/.git/worktrees/invapi-php-store
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

None, direct invocation of relevant .exe file.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git status
```

 - What did you expect to occur after running these commands?

```
On branch DEVOPS-889
Your branch is up to date with 'origin/DEVOPS-889'.
```

 - What actually happened instead?

`fatal: not a git repository (or any of the parent directories): .git`

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Easily reproducible with any given repository.

Cygwin Git has no issues with such layout. But Git for Windows, somehow, fails.

**Comment by dscho** [comment]
> * What commands did you run to trigger this issue? If you can provide a
>       [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
>       this will help us understand the issue.
> 
> 
> ```
> git status
> ```

This is not a verifiable example:

```console
$ git status
fatal: not a git repository (or any of the parent directories): .git
```

**Comment by AnrDaemon** [comment]
Here's a simple reproducer for you.

```cmd
@ECHO OFF
SET PATH=C:\Programs\Git\bin;%PATH%
git init wtest
cd wtest
echo "Worktree test" > README.md
git add README.md
git commit -m "Initial commit"
git worktree add -B wtest-1 ../wtest-1 master
cd ../wtest-1
ATTRIB -H -S -R .git
RENAME .git .git~
MKLINK /J .git ..\wtest\.git\worktrees\wtest-1
git status
```

Adjust `%PATH%` to your GfW installation and run.
Then try `git status` in the worktree dir using Cygwin Git. It just works.

```
~\Documents\git-projects> wtest.cmd
Initialized empty Git repository in ~/Documents/git-projects/wtest/.git/
[master (root-commit) 27e15ca] Initial commit
 1 file changed, 1 insertion(+)
 create mode 100644 README.md
Preparing worktree (new branch 'wtest-1')
HEAD is now at 27e15ca Initial commit
соединение создано для .git <<===>> ..\wtest\.git\worktrees\wtest-1
fatal: not a git repository (or any of the parent directories): .git

~\Documents\git-projects\wtest-1> "%CYGWIN_DIR%\bin\git" status
On branch wtest-1
nothing to commit, working tree clean
```

**Comment by dscho** [comment]
I could reproduce, but I had to change the forward slash into a backslash. This is my Batch file:

```cmd
git init wtest
git -C wtest commit -m initial --allow-empty
git -C wtest worktree add -B wtest-1 ../wtest-1 HEAD
cd wtest-1
ATTRIB -H -S -R .git
RENAME .git .git~
REM MKLINK /J .git ..\wtest\.git\worktrees\wtest-1
MKLINK /D .git ..\wtest\.git\worktrees\wtest-1
dir
git status
```

The reason is that when Git reads and interprets the `...\wtest-1\.git\commondir` file (whose contents are `../..`), it concatenates the contents to the path of the containing directory (`...\wtest-1\.git\../..`) , and then uses [`GetFinalPathNameByHandleW()`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfinalpathnamebyhandlea?redirectedfrom=MSDN) to resolve the `..` components, getting it wrong because it simply strips `../..` and the two components preceding that.

To show what I mean, I wrote this little C program:

```c
#include <stdio.h>
#include <windows.h>

int main(int argc, char **argv)
{
	const char *path = argc > 1 ? argv[1] : ".";
	HANDLE h = CreateFileA(path, 0,
			FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
			OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
	char buffer[MAX_PATH];
	DWORD ret;

	strncpy(buffer, path, sizeof(buffer));
	buffer[sizeof(buffer) - 1] = '\0';
	ret = GetFinalPathNameByHandleA(h, buffer, sizeof(buffer), 0);
	CloseHandle(h);
	if (!ret || ret > sizeof(buffer)) {
		fprintf(stderr, "There was an error: %ld\n", GetLastError());
		return 1;
	} else {
		printf("%s\n", buffer);
		return 0;
	}
}
```

If you compile it and run it with the single parameter `wtest-1/.git/../..`, it will print the absolute path of the current directory. With `wtest-1/.git/..`, it will print the absolute path of the `wtest-1` directory, but with `wtest-1/.git` it will print the absolute path to `wtest/.git/worktrees/wtest-1`.

Bottom line: keep using the `.git` _file_, and do not replace it with an NTFS junction, that will only cause problems.

**Comment by AnrDaemon** [comment]
It actually solves problem. Not every tool is happy to run executables to obtain simple information about current repository. Reading from `~/.git/HEAD` is multiple times faster than running `git branch --show-current`, and safer, too.

However, it exposes another problem in how Git resolves paths.

**Comment by AnrDaemon** [comment]
What I don't understand is why Cygwin Git works while Windows one does not. (I can't use Cygwin Git with native Windows tools, unfortunately, else I wouldn't even notice this issue.)

**Comment by dscho** [comment]
> Reading from `~/.git/HEAD` is multiple times faster than running `git branch --show-current`,

Yes. But both are relatively fast. There are performance issues in Git where improvements would have a much larger impact.

> and safer,

No. There is currently work going on to support refs backends other than the `files` one. Once that work is complete, you may very well end up with broken tools if you read `.git/HEAD` directly. As a rule of thumb, Git considers pretty much everything inside `.git` as implementation detail, subject to change, and if you rely on it being in a certain format, you set yourself up for failure. I would highly recommend to stop accessing files within `.git/` altogether, and use the Git executable exclusively to interact with the contents of that folder.
